pkg/api/mlflow/dao/repositories/param.go

```diff
@@ -32,48 +43,65 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
 
 // CreateBatch creates []models.Param entities in batch.
 func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params []models.Param) error {
-	// try to create params in batch; error condition requires special handling
-	// to allow certain duplicates
-	if err := r.db.CreateInBatches(params, batchSize).Error; err != nil {
-		// remove duplicate rows and try again
-		dedupedParams, errRemovingMatches := r.removeExactMatches(ctx, params)
-		if errRemovingMatches != nil {
-			return eris.Wrap(errRemovingMatches, "error removing duplicates in batch")
+	tx := r.db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "run_uuid"}, {Name: "key"}},
+		DoNothing: true,
+	}).CreateInBatches(params, batchSize)
+	if tx.Error != nil {
+		return eris.Wrap(tx.Error, "error creating params in batch")
+	}
+
+	// if there are conflicting params, ignore if the values are the same
+	if tx.RowsAffected != int64(len(params)) {
+		conflictingParams, err := r.findConflictingParams(ctx, params)
+		if err != nil {
+			return eris.Wrap(err, "error checking for conflicting params")
 		}
-		if err := r.db.CreateInBatches(dedupedParams, batchSize).Error; err != nil {
-			return eris.Wrap(err, "error creating params in batch after removing duplicates")
+		if len(conflictingParams) > 0 {
+			return ParamConflictError{
+				Message: fmt.Sprintf("conflicting params found: %v", conflictingParams),
+			}
 		}
 	}
 	return nil
 }
 
-// removeExactMatches will return a new slice of params which excludes exact matches
-func (r ParamRepository) removeExactMatches(_ context.Context, params []models.Param) ([]models.Param, error) {
-	var keys []string
-	paramMap := map[string]models.Param{}
-	for _, param := range params {
-		key := fmt.Sprintf("%v-%v-%v", param.RunID, param.Key, param.Value)
-		keys = append(keys, key)
-		paramMap[key] = param
+// findConflictingParams checks if there are conflicting values for the input params. If a key does not
+// yet exist in the db, or if the same key and value already exist for the run, it is not a conflict.
+// If the key already exists for the run but with a different value, it is a conflict. Conflicting keys are returned.
+func (r ParamRepository) findConflictingParams(ctx context.Context, params []models.Param) ([]string, error) {
```

Comment:
this whole function could be changed by a single SQL query in this spirit:
```sql
WITH new(key, value) AS (VALUES (?, ?), (?, ?))
SELECT current.key, current.value, new.value
FROM params AS current
INNER JOIN new ON run_uuid = ? AND new.key = current.key
WHERE new.value != current.value;
```

`VALUES` can be built as a list of  `n` `"(?, ?)"` joined by `", "`, where `n` is the amount of params we tried to insert.

The SQL parameters that need to be passed are the list keys and values we tried to insert, and then the run ID.

The DB engine will then return to you all you need to log the error nicely to the user.

Reply:
I can change to this, but I'd suggest a simple query like we have currently plus Go-layer manipulation is easier to grok